File: cpp/questions/permute.cpp
#include <functional>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        int ns = nums.size();
        std::vector<std::vector<int>> ans;
        std::vector<bool> flag(ns);        
        std::vector<int> tmp(ns);
        std::function<void(int)> bv = [&](int p) -> void {
            if (p == ns) {
                ans.emplace_back(tmp);
            }
            for (int i=0; i<ns; ++i) {
                if (!flag[i]) {
                    tmp[p] = nums[i];
                    flag[i] = true;
                    bv(p+1);
                    flag[i] = false;
                }
            }
        };
        bv(0);
        return ans;
    }
};

File: python/questions/permute.py
from typing import List

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ln = len(nums)
        ans = []
        fg = [False] * ln
        tmp = [0] * ln
        def bv(p: int) -> None:
            nonlocal tmp, fg, nums, ans, ln
            if p == ln:
                ans.append(tmp.copy())
            for i in range(ln):
                if not fg[i]:
                    tmp[p] = nums[i]
                    fg[i] = True
                    bv(p+1);
                    fg[i] = False
        bv(0)
        return ans
